chain-abci/src/app/app_init.rs

```diff
@@ -466,20 +490,68 @@ impl<T: EnclaveProxy> ChainNodeApp<T> {
                 network_params,
                 validator_state,
             );
-            self.storage
-                .store_genesis_state(&genesis_state, self.tx_query_address.is_some());
+            chain_storage::store_genesis_state(
+                &mut kv_store!(self),
+                &genesis_state,
+                self.tx_query_address.is_some(),
+            );
 
-            let wr = self.storage.persist_write();
-            if let Err(e) = wr {
-                panic!("db write error: {}", e);
-            } else {
-                self.uncommitted_account_root_hash = genesis_state.top_level.account_root;
-                self.last_state = Some(genesis_state);
-            }
+            flush_storage(&mut self.storage, std::mem::take(&mut self.kv_buffer)).unwrap();
+            self.last_state = Some(genesis_state);
 
             ResponseInitChain::new()
         } else {
             panic!("validators in genesis configuration are not consistent with app_state")
         }
     }
+
+    pub fn staking_store(&mut self, deliver: bool) -> impl StoreStaking + '_ {
+        let root = self
+            .last_state
+            .as_ref()
+            .map(|state| state.top_level.account_root);
+        staking_store!(self, root, deliver)
+    }
+
+    pub fn staking_getter(&self, deliver: bool) -> impl GetStaking + '_ {
+        let root = self
+            .last_state
+            .as_ref()
+            .map(|state| state.top_level.account_root);
+        staking_getter!(self, root, deliver)
+    }
+
+    pub fn kv_store(&mut self, deliver: bool) -> impl StoreKV + '_ {
```

Comment:
it's unclear what this `deliver: bool` param is meant to be set when. As these are public functions, it'll be good to document them -- either with top-level comments, or using a more descriptive `enum` type instead of bool

Reply:
Done, Changed to `BufferType`.